python/setup_project/storage_map_page.py

```diff
@@ -0,0 +1,398 @@
+# Copyright (c) 2018 Shotgun Software Inc.
+#
+# CONFIDENTIAL AND PROPRIETARY
+#
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+import os
+import sys
+import traceback
+
+import sgtk
+from sgtk.platform.qt import QtCore
+from sgtk.platform.qt import QtGui
+from sgtk.util.filesystem import ensure_folder_exists
+from sgtk.util.storage_roots import StorageRoots
+
+from .base_page import BasePage
+from .storage_map_widget import StorageMapWidget
+
+logger = sgtk.platform.get_logger(__name__)
+
+# use sg utils settings to remember previous storage mappings
+settings = sgtk.platform.import_framework(
+    "tk-framework-shotgunutils",
+    "settings"
+)
+
+
+class StorageMapContainerWidget(QtGui.QWidget):
+    """
+    This class defines the widget contained within the scroll area where the
+    mapping widgets will be added.
+
+    The primary purpose of this class is to add a layout to the widget so
+    that the mapping widgets can be added to it.
+    """
+
+    def __init__(self, parent=None):
+        super(StorageMapContainerWidget, self).__init__(parent)
+
+        layout = QtGui.QVBoxLayout()
+        layout.setSpacing(6)
+        layout.setContentsMargins(6, 6, 6, 6)
+
+        self.setLayout(layout)
+
+
+class StorageModel(QtGui.QStandardItemModel):
+    """
+    A simple modle to store all available storages.
+
+    Some storages may exists in SG, others may be created/edited/saved via the
+    storage map widgets. The model is shared by all map widgets to keep them
+    in sync as storages are created/updated.
+    """
+
+    # non-storage item text
+    CHOOSE_STORAGE_ITEM_TEXT = "Choose..."
+    CREATE_STORAGE_ITEM_TEXT = "+ New"
+
+    # data role for the storages in the model's items
+    STORAGE_DATA_ROLE = QtCore.Qt.UserRole + 7
+
+    def __init__(self, parent):
+        """Initialize the model."""
+
+        super(StorageModel, self).__init__(parent)
+
+        # query all existing SG storages to include in the model
+        sg_connection = sgtk.platform.current_engine().shotgun
```

Comment:
Stream of consciousness notes: There are special rules for uniqueness for local storages - they have to have unique names across both active and retired storages! If you haven't arlready, can you do a quick test to check that if we try to create a new storage named the same thing as a retired storage, the error message bubbles up to the user and makes sense?

Reply:
As discussed offline, it's not possible to simply retire a local storage in SG these days. Will leave as is since it's likely an edge case that clients will have old/retired storages.